lib/cn/cn.c

```diff
@@ -773,28 +773,63 @@ cn_maint_task(struct work_struct *work)
 }
 
 struct cndb_cn_ctx {
-    struct cn  *cn;
-    struct map *nodemap;
+    struct cn_tree *tree;
+    struct map     *nodemap;
 };
 
-/* This function is the cndb_cn_instantiate() callback.  It is called
- * once for each kvset found in the cndb for the cn specified in the
- * callback context.
+static merr_t
+cndb_cn_ctx_init(struct cndb_cn_ctx *ctx, struct cn_tree *tree, struct cn_tree_node *root)
+{
+    struct map *nodemap;
+
+    INVARIANT(ctx);
+    INVARIANT(tree);
+    INVARIANT(root);
+
+    nodemap = map_create(CN_FANOUT_MAX);
+    if (!nodemap)
+        return merr(ENOMEM);
+
+    ctx->nodemap = nodemap;
+    ctx->tree = tree;
+
+    /* Pre-populate node map with root node, which always has ID 0. */
+    map_insert_ptr(ctx->nodemap, 0, root);
+
+    return 0;
+}
+
+static void
+cndb_cn_ctx_fini(struct cndb_cn_ctx *ctx)
+{
+    INVARIANT(ctx);
+    INVARIANT(ctx->nodemap);
+
+    map_destroy(ctx->nodemap);
+}
+
+/**
+ * Callback invoked by cndb_cn_instantiate() to place kvsets into tree nodes.
+ *
+ * This callback is invoked once for each kvset in a KVS.  Each callback
+ * contains a node ID, a kvset ID, and other metadata needed to open the
+ * on-media kvset.  It opens the kvsets and adds them to tree nodes, creating
+ * the tree nodes as needed.
  */
 static merr_t
-cn_kvset_cb(void *arg, struct kvset_meta *km, u64 kvsetid)
+cndb_cn_callback(void *arg, struct kvset_meta *km, u64 kvsetid)
 {
     struct cndb_cn_ctx *ctx = arg;
     struct cn_tree_node *node;
     struct cn_tree *tree;
     struct kvset *kvset;
     merr_t err;
 
-    tree = ctx->cn->cn_tree;
+    tree = ctx->tree;
 
     node = map_lookup_ptr(ctx->nodemap, km->km_nodeid);
     if (!node) {
-        node = cn_node_alloc(tree, km->km_nodeid);
+        node = cn_node_alloc(ctx->tree, km->km_nodeid);
```

Comment:
why not just "tree" ?

Reply:
I started to get rid of local var tree, but then backed out.  Will proceed with getting rid of it.

Reply:
We could probably have just added a function to the route map to find an entry by node ID, then we could just get rid of all this map stuff...

Reply:
This way we build nodes, then create route map / tree.  Which is kind of nice.  Baby stepping our way through cn_open.

Reply:
Yeah, we had a bug where I didn't wait for all the kvsets to be added before looking for the max key, so we kinda have to do it this way.  I suppose you could update the max key on each kvset but that'd be silly..